### PR TITLE
 ✨ Add vcsim server provider

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds a vcsim server that can be used in tilt for:
- Testing CAPV in legacy mode, using vcsim instead of a real vCenter
- Testing CAPV in supervisor mode, using vcsim instead of a real vCenter

The second option also provides some of the building blocks for implementing E2E for CAPV in supervisor mode on VMC (or on any vSphere), but some work is still necessary to make this possible

### CODE REVIEW

**Comment by fabriziopandini** [comment]
/hold
It seems there is ongoing work in vcsim that could help us here, I'm gathering info

**Comment by sbueringer** [comment]
@fabriziopandini IIRC you wanted to make some further changes before review, correct?

**Comment by fabriziopandini** [comment]
/hold cancel
the ongoing work in vcsim will make it easier some parts of the CAPV testing effort, but we still need this work if we want to keep the simulated environment as lightweight as possible (e.g. don't use a container for each VM)

**Comment by fabriziopandini** [comment]
@sbueringer this one should be ready for review now

**Comment by sbueringer** [comment]
/test pull-cluster-api-provider-vsphere-e2e-main

(to run all tests)

File: test/infrastructure/vcsim/controllers/controlplaneendpoint_controller.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		return ctrl.Result{}, err
+	}
+
+	// Always attempt to Patch the controlPlaneEndpoint object and status after each reconciliation.
+	defer func() {
+		if err := patchHelper.Patch(ctx, controlPlaneEndpoint); err != nil {
+			log.Error(err, "failed to patch ControlPlaneEndpoint")
+			if reterr == nil {
+				reterr = err
+			}
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/controlplaneendpoint_controller.go:75]:
```suggestion
			reterr = kerrors.NewAggregate([]error{reterr, err})
```

(we recently changed this in ~ all our controllers)

File: test/infrastructure/vcsim/controllers/envsubst_controller.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	vCenter := &vcsimv1.VCenter{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: envSubst.Namespace,
+		Name:      envSubst.Spec.VCenter,
+	}, vCenter); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get VCenter")
+	}
+	log = log.WithValues("VCenter", klog.KObj(vCenter))
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/envsubst_controller.go:81]:
```suggestion
	log = log.WithValues("VCenter", klog.KObj(vCenter))
	ctx = ctrl.LoggerInto(ctx, log)
```

Let's directly update the ctx. So the logger is already up-to-date wherever we use the ctx

File: test/infrastructure/vcsim/controllers/controlplaneendpoint_controller.go (hunk 1 above)
Comment (sbueringer) [reply test/infrastructure/vcsim/controllers/controlplaneendpoint_controller.go:75]: Same in other controllers

File: test/infrastructure/vcsim/controllers/envsubst_controller.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		if err != nil {
+			return errors.Wrapf(err, "failed to generate public key")
+		}
+
+		sshKey = string(publicKeyBytes)
+		r.sshKeys[key] = sshKey
+		log.Info("Created ssh authorized key")
+	}
+
+	// Common variables (used both in supervisor and legacy mode)
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/envsubst_controller.go:155]:
```suggestion
	// Common variables (used both in supervisor and govmomi mode)
```
(similar in other cases)

File: test/infrastructure/vcsim/controllers/images/images.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+// Package images contains fake images for the vcsim content library.
+package images
+
+import (
+	_ "embed"
+)
+
+var (
+	// SampleOVF image.
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/images/images.go:25]: Do we have to use "real" files or would ~ empty ones also work?

File: test/infrastructure/vcsim/controllers/vcenter_controller.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+		}
+	}
+
+	if vCenter.Status.Thumbprint == "" {
+		// Compute the Thumbprint out of the certificate self-generated by vcsim.
+		config := &tls.Config{InsecureSkipVerify: true} //nolint: gosec
+		addr := vCenter.Status.Host
+		conn, err := tls.Dial("tcp", addr, config)
+		if err != nil {
+			return errors.Wrapf(err, "failed to connect to VCSim Server instance to infert thumbprint")
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vcenter_controller.go:213]:
```suggestion
			return errors.Wrapf(err, "failed to connect to VCSim Server instance to infer thumbprint")
```

File: test/infrastructure/vcsim/controllers/vcenter_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	g.Expect(err).ToNot(HaveOccurred())
+
+	res, err = r.Reconcile(ctx, ctrl.Request{NamespacedName: types.NamespacedName{
+		Namespace: vCenter.Namespace,
+		Name:      vCenter.Name,
+	}})
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(res).To(Equal(ctrl.Result{}))
+
+	// Check the VCenter instance has been created in the reconciler internal status
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vcenter_controller_test.go:109]:
```suggestion
	// Check the VCenter instance has been deleted in the reconciler internal status
```

File: test/infrastructure/vcsim/controllers/vcsim.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	//  we create a template in a datastore, what if many?
+	//  we create a template in a cluster, but the generated vm doesn't have the cluster in the path. What if I have many clusters?
+	cluster := 0
+	datastore := 0
+	datacenters := 1
+	if vCenter.Spec.Model != nil {
+		datacenters = ptr.Deref(vCenter.Spec.Model.Datacenter, simulator.VPX().Datacenter) // VPX is the same base model used when creating vcsim
+	}
+	for dc := 0; dc < datacenters; dc++ {
+		exit := cli.Run([]string{"vm.create", fmt.Sprintf("-ds=%s", vcsimDatastoreName(datastore)), fmt.Sprintf("-cluster=%s", vcsimClusterName(dc, cluster)), fmt.Sprintf("-net=%s", vcsimDefaultNetworkName), "-disk=20G", "-on=false", "-k=true", fmt.Sprintf("-u=%s", govcURL), vcsimDefaultVMTemplateName})
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vcsim.go:90]:
```suggestion
		exit := cli.Run([]string{"vm.create", fmt.Sprintf("-ds=%s", vcsimDatastoreName(datastore)), fmt.Sprintf("-cluster=%s", vcsimClusterName(dc, cluster)), fmt.Sprintf("-net=%q", vcsimDefaultNetworkName), "-disk=20G", "-on=false", "-k=true", fmt.Sprintf("-u=%s", govcURL), vcsimDefaultVMTemplateName})
```

Fine if it works. Wondering if we don't have to quote the string as it has spaces

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+			return virtualMachine.Status.PowerState == vmoprv1.VirtualMachinePoweredOn && virtualMachine.Status.VmIp != "" && virtualMachine.Status.BiosUUID != ""
+		},
+		GetProviderID: func() string {
+			// Computes the ProviderID for the node hosted on the virtualMachine
+			return util.ConvertUUIDToProviderID(virtualMachine.Status.BiosUUID)
+		},
+	}
+}
+
+func (r VirtualMachineReconciler) getVCenterSession(ctx context.Context) (*session.Session, error) {
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:270]:
```suggestion
func (r *VirtualMachineReconciler) getVCenterSession(ctx context.Context) (*session.Session, error) {
```

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	// Fetch the owner VSphereMachine.
+	vSphereMachine, err := util.GetOwnerVMWareMachine(ctx, r.Client, virtualMachine.ObjectMeta)
+	// vsphereMachine can be nil in cases where custom mover other than clusterctl
+	// moves the resources without ownerreferences set
+	// in that case nil vsphereMachine can cause panic and CrashLoopBackOff the pod
+	// preventing vspheremachine_controller from setting the ownerref
+	if err != nil || vSphereMachine == nil {
+		log.Info("Owner VSphereMachine not found, won't reconcile", "key", req.NamespacedName)
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:88]:
```suggestion
		log.Info("Owner VSphereMachine not found, won't reconcile")
```

We already have the VirtualMachine in the logger

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+		return reconcile.Result{}, nil
+	}
+	log = log.WithValues("VSphereMachine", klog.KObj(vSphereMachine))
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:91]:
```suggestion
	log = log.WithValues("VSphereMachine", klog.KObj(vSphereMachine))
	ctx = ctrl.LoggerInto(ctx, log)
```

Similar below. Otherwise we use a slightly outdated logger in the util funcs

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+	// Fetch the owner VSphereMachine.
+	vSphereMachine, err := util.GetOwnerVMWareMachine(ctx, r.Client, virtualMachine.ObjectMeta)
+	// vsphereMachine can be nil in cases where custom mover other than clusterctl
+	// moves the resources without ownerreferences set
+	// in that case nil vsphereMachine can cause panic and CrashLoopBackOff the pod
+	// preventing vspheremachine_controller from setting the ownerref
+	if err != nil || vSphereMachine == nil {
+		log.Info("Owner VSphereMachine not found, won't reconcile", "key", req.NamespacedName)
+		return reconcile.Result{}, nil
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:89]:
```suggestion
		return reconcile.Result{}, err
```
I think?

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	// Always attempt to Patch the VSphereVM + conditionsTracker object and status after each reconciliation.
+	defer func() {
+		// NOTE: Patch on VSphereVM will only add/remove a finalizer.
+		if err := patchHelper.Patch(ctx, virtualMachine); err != nil {
+			log.Error(err, "failed to patch VSphereVM")
+			if reterr == nil {
+				reterr = err
+			}
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:175]:
```suggestion
			reterr = kerrors.NewAggregate([]error{reterr, err})
```

Same below

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+		return ret, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *VirtualMachineReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, virtualMachine *vmoprv1.VirtualMachine, conditionsTracker *infrav1.VSphereVM) error {
+	bootstrapReconciler := r.getVMBootstrapReconciler(virtualMachine)
+	if ret, err := bootstrapReconciler.reconcileShutdown(ctx, cluster, machine, conditionsTracker); !ret.IsZero() || err != nil {
+		return err
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:220]:
```suggestion
		return ret. err
```
Otherwise we just ignore !ret.IsZero

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+	// Compute the resource group unique name.
+	// NOTE: We are using reconcilerGroup also as a name for the listener for sake of simplicity.
+	resourceGroup := klog.KObj(cluster).String()
+
+	// Check if there is a conditionsTracker in the resource group.
+	// The conditionsTracker is an object stored in memory with the scope of storing conditions used for keeping
+	// track of the provisioning process of the fake node, etcd, api server, etc for this specific virtualMachine.
+	// (the process managed by this controller).
+	inmemoryClient := r.InMemoryManager.GetResourceGroup(resourceGroup).GetClient()
+	conditionsTracker := &infrav1.VSphereVM{} // NOTE: The type of this object doesn't matter as soon as it implements Cluster API's conditions interfaces.
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/virtualmachine_controller.go:145]: Wondering if using something else would be less confusing :) (supervisor vs. infrav1.VSphereVM)

E.g. a vmoprv1.VirtualMachine{}

(or maybe define a separate struct)

File: test/infrastructure/vcsim/controllers/vspherevm_controller.go [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the owner VSphereMachine.
+	vSphereMachine, err := util.GetOwnerVSphereMachine(ctx, r.Client, vSphereVM.ObjectMeta)
+	// vsphereMachine can be nil in cases where custom mover other than clusterctl
+	// moves the resources without ownerreferences set
+	// in that case nil vsphereMachine can cause panic and CrashLoopBackOff the pod
+	// preventing vspheremachine_controller from setting the ownerref
+	if err != nil || vSphereMachine == nil {
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vspherevm_controller.go:90]: Similar comments as for virtualmachine_controller.go please compare

File: test/infrastructure/vcsim/controllers/vmip.go [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+	log.Info("Customizing the VM for applying an IP")
+	task, err = vm.Customize(ctx, spec)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to Customize vm")
+	}
+	if err = task.Wait(ctx); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to wait for Customize vm task to complete")
+	}
+
+	log.Info("Powering On the VM before applying an IP")
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vmip.go:124]:
```suggestion
	log.Info("Powering On the VM after applying the IP")
```
?

File: test/infrastructure/vcsim/controllers/vmbootstrap.go [hunk 17]
```diff
@@ -0,0 +1,1002 @@
...
+
+		// NOTE: for the first control plane machine we might create the node before etcd and API server pod are running
+		// but this is not an issue, because it won't be visible to CAPI until the API server start serving requests.
+		if err := inmemoryClient.Create(ctx, node); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create Node")
+		}
+	}
+
+	conditions.MarkTrue(conditionsTracker, NodeProvisionedCondition)
+	log.Info("Node created", "Pod", klog.KObj(node))
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vmbootstrap.go:264]:
```suggestion
	log.Info("Node created", "Node", klog.KObj(node))
```
I think

File: test/infrastructure/vcsim/controllers/vmbootstrap.go [hunk 18]
```diff
@@ -0,0 +1,1002 @@
...
+		}
+
+		if err := inmemoryClient.Create(ctx, corednsDeployment); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create coreDNS deployment")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *vmBootstrapReconciler) reconcileShutdown(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, conditionsTracker *infrav1.VSphereVM) (ctrl.Result, error) {
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vmbootstrap.go:790]: We are we calling this here all shutdown while it's delete in core CAPI. Is there a semantic difference?

**Review by sbueringer — COMMENTED** [review]
I changed my mind and added comments instead of opening a PR (A large percentag of my points are questions)

File: tilt-provider.json [hunk 19]
```diff
@@ -1,18 +1,38 @@
...
+    "name": "vsphere-supervisor",
+    "config": {
+      "version": "v1.10.99",
+      "image": "gcr.io/k8s-staging-capi-vsphere/cluster-api-vsphere-controller",
+      "live_reload_deps": [
+        "main.go",
+        "go.mod",
+        "go.sum",
+        "api",
+        "contrib",
```
Comment (chrischdi) [inline tilt-provider.json:30]:
```suggestion
        "apis",
```

File: test/framework/vmoperator/vmoperator.go [hunk 20]
```diff
@@ -0,0 +1,576 @@
...
+		for i := range libraries {
+			if libraries[i].Name == contentLibrary.Name {
+				contentLibraryID = libraries[i].ID
+				break
+			}
+		}
+	}
+
+	if contentLibraryID == "" {
+		id, err := libMgr.CreateLibrary(ctx, contentLibrary)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/framework/vmoperator/vmoperator.go:397]: Good catch

File: test/infrastructure/vcsim/main.go [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+func setupWebhooks(_ ctrl.Manager, _ bool) {
+}
+
+func concurrency(c int) controller.Options {
+	return controller.Options{MaxConcurrentReconciles: c}
+}
+
+func isSupervisor(mgr ctrl.Manager) (bool, error) {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/infrastructure/vcsim/main.go:372]: Maybe let's do the same as in capv

File: tilt-provider.json [hunk 22]
```diff
@@ -1,18 +1,38 @@
...
+        "controllers",
+        "pkg"
+      ],
+      "label": "CAPV",
```
Comment (sbueringer) [inline tilt-provider.json:34]: We might want a different label that doesn't conflict. But not sure

Comment (sbueringer) [reply tilt-provider.json:34]: Maybe this also blocks deploying both at the same time, which would be probably good

File: tilt-provider.json [hunk 23]
```diff
@@ -1,18 +1,38 @@
...
+      ],
+      "label": "CAPV"
+    }
+  },
+  {
+    "name": "vsphere-supervisor",
+    "config": {
+      "version": "v1.10.99",
+      "image": "gcr.io/k8s-staging-capi-vsphere/cluster-api-vsphere-controller",
+      "live_reload_deps": [
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply tilt-provider.json:25]: fixed

File: test/framework/vmoperator/vmoperator.go [hunk 24]
```diff
@@ -0,0 +1,576 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"sigs.k8s.io/cluster-api-provider-vsphere/packaging/flavorgen/flavors/util"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/session"
+)
+
+const DefaultNamespace = "vmware-system-vmop"
+
+const (
+	// copied from vm-operator.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/vmoperator/vmoperator.go:51]: we copied only const (also added to the comment)

File: test/framework/vmoperator/vmoperator.go [hunk 25]
```diff
@@ -0,0 +1,576 @@
...
+	providerConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      providerConfigMapName,
+			Namespace: config.Namespace,
+		},
+		Data: map[string]string{
+			caFilePathKey: "", // Leaving this empty because we don't have (yet) a solution to inject a CA file into the vm-operator pod.
+			// Cluster exists in the example ConfigMap, but it is not defined as a const
+			// ContentSource exists in the example ConfigMap, but it is not defined as a const
+			// CtrlVmVmAATag exists in the example ConfigMap, but it is not defined as a const
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/vmoperator/vmoperator.go:288]: simplified all the comments, those were mostly my notes why figuring out the minimal set of required values by comparing with an example ConfigMap grabbed from a real supervisor cluster

File: test/framework/vmoperator/vmoperator.go [hunk 26]
```diff
@@ -0,0 +1,576 @@
...
+			return errors.Wrapf(err, "failed to create vm-operator VirtualMachineClassBinding %s", vmClassBinding.Name)
+		}
+		log.Info("Created vm-operator VirtualMachineClassBinding", "VirtualMachineClassBinding", klog.KObj(vmClassBinding))
+	}
+
+	// Create a ContentLibrary in K8s and in vCenter, bind it to the K8s namespace
+	// This requires a set of objects in vc-sim as well as their mapping in K8s
+	// - vcsim: a Library containing an Item
+	// - k8s: ContentLibraryProvider, ContentSource (both representing the library), a VirtualMachineImage (representing the Item)
+
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/vmoperator/vmoperator.go:363]: also a real vCenter; fixed in the entire file

File: test/framework/vmoperator/vmoperator.go [hunk 27]
```diff
@@ -0,0 +1,576 @@
...
+		if err != nil {
+			return errors.Wrapf(err, "failed to start update session for vm-operator libraryItem %s", libraryItem.Name)
+		}
+		if err := uploadFunc(sessionID, file.Name, file.Content); err != nil {
+			return errors.Wrapf(err, "failed to upload data for vm-operator libraryItem %s", libraryItem.Name)
+		}
+		if err := libMgr.CompleteLibraryItemUpdateSession(ctx, sessionID); err != nil {
+			return errors.Wrapf(err, "failed to complete update session for vm-operator libraryItem %s", libraryItem.Name)
+		}
+		log.Info("Uploaded vm-operator LibraryItemFile in vcsim", "ContentLibrary", contentLibrary.Name, "libraryItem", klog.KObj(virtualMachineImage), "LibraryItemFile", file.Name)
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/vmoperator/vmoperator.go:573]: fixed

File: test/infrastructure/vcsim/Dockerfile [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+RUN --mount=type=cache,target=/go/pkg/mod \
+    go mod download
+
+# This needs to build with the entire Cluster API context
+WORKDIR /workspace
+# Copy the sources (which includes the test/infrastructure/inmemory subdirectory)
+COPY ./ ./
+
+# Essentially, change directories into CAPIM
+WORKDIR /workspace/test/infrastructure/inmemory
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/Dockerfile:56]: Fixed + added the make target, which will be used when we integrate this with CI

File: test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+
+// VCCenterSpec defines the desired state of the VCenter.
+type VCCenterSpec struct {
+	Model *VCSimModelSpec `json:"model,omitempty"`
+}
+
+// VCSimModelSpec defines the model to be used by the VCenter.
+type VCSimModelSpec struct {
+	// VSphereVersion specifies the VSphere version to use
+	// Default: 7.0.0
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go:37]: This is the minimal vCenter version required by CAPV (vcsim default is 6.5)

File: test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+	// Name prefix: H, vcsim flag: -host
+	// Default: 3
+	ClusterHost *int `json:"clusterHost,omitempty"`
+
+	// Pool specifies the number of ResourcePool entities to create per Cluster
+	// Note that every cluster has a root ResourcePool named "Resources", as real vCenter does.
+	// For example: /DC0/host/DC0_C0/Resources
+	// The root ResourcePool is named "RP0" within other object names.
+	// When Model.Pool is set to 1 or higher, this creates child ResourcePools under the root pool.
+	// Note that this flag is not effective on standalone hosts.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go:60]: standalone hosts = ESXi without vCenter in this context (this is copied from vcsim model types).

File: test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+	// For example: /DC0/host/DC0_C0/Resources/DC0_C0_RP1
+	// Name prefix: RP, vcsim flag: -pool
+	// Default: 0
+	// TODO: model pool selection for each cluster; for now ResourcePool named "Resources" will be always used
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go:64]: Added more context in the TODO
TL;DR; vcsim has a resource pool named "resources" and according to documentation it can create more named RP1, RP2 etc. This TODO is about allowing to select RPx node pools for a specific cluster

File: test/infrastructure/vcsim/config/manager/manager.yaml [hunk 32]
```diff
@@ -0,0 +1,46 @@
+apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/config/manager/manager.yaml:1]: Fixed (missed this when starting adding controllers)

File: test/infrastructure/vcsim/controllers/envsubst_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply test/infrastructure/vcsim/controllers/envsubst_controller.go:155]: Fixed everywhere

File: tilt-provider.json (hunk 22 above)
Comment (fabriziopandini) [reply tilt-provider.json:34]: It is not possible to deploy both at the same time

File: test/framework/vmoperator/vmoperator.go [hunk 33]
```diff
@@ -0,0 +1,576 @@
...
+			return errors.Wrapf(err, "failed to get vm-operator StorageClass %s", storageClass.Name)
+		}
+
+		if err := c.Create(ctx, storageClass); err != nil {
+			return errors.Wrapf(err, "failed to create vm-operator StorageClass %s", storageClass.Name)
+		}
+		log.Info("Created vm-operator StorageClass", "StorageClass", klog.KObj(storageClass))
+	}
+
+	// TODO: rethink about this, for now we are creating a ResourceQuota with the same name of the StorageClass, might be this is not ok when hooking into a real vCenter
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/framework/vmoperator/vmoperator.go:193]: Maybe let's add ~ that as the comment "It seem like vm-operator is using the ResourceQuota to figure out which StorageClass can be used from a namespace"

Better than nothing :)

File: test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/infrastructure/vcsim/api/v1alpha1/vcenter_types.go:37]: Might be worth a quick comment. So we have some data when we want to re-evaluate later. But also fine for me to just resolve

File: test/infrastructure/vcsim/main.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/main.go:372]: For now, I just aligned the two main ones, we can change it as a follow-up (it is kind of a breaking change)

File: test/infrastructure/vcsim/controllers/vmbootstrap.go [hunk 34]
```diff
@@ -0,0 +1,1002 @@
...
+type vmBootstrapReconciler struct {
+	Client          client.Client
+	InMemoryManager inmemoryruntime.Manager
+	APIServerMux    *inmemoryserver.WorkloadClustersMux
+
+	IsVMReady     func() bool
+	GetProviderID func() string
+}
+
+func (r *vmBootstrapReconciler) reconcileBoostrap(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, conditionsTracker *infrav1.VSphereVM) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/vcsim/controllers/vmbootstrap.go:148]: Nice idea, done!

File: test/infrastructure/vcsim/controllers/virtualmachine_controller.go (hunk 14 above)
Comment (fabriziopandini) [reply test/infrastructure/vcsim/controllers/virtualmachine_controller.go:145]: improved the comment
unfortunately vmoprv1.VirtualMachine isn't a condition getter

File: test/infrastructure/vcsim/controllers/images/images.go (hunk 4 above)
Comment (fabriziopandini) [reply test/infrastructure/vcsim/controllers/images/images.go:25]: Dropped the biggest one, but I would leave the smallest one so we have something testing the code that uploads files to a content library.
Is that ok?

File: test/framework/vmoperator/vmoperator.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/vmoperator/vmoperator.go:397]: This func would create the files that you pass in input in the config Dependencies; for vcsim we are passing dummy files in the VCenter controller, but nothing prevents us from passing real OVA or no files at all.

For VMC or other environments where a content library already exists, we can pass the library name in the config, and the code will detect that it already exists, the same for the OVA (and you can omit files in this case).
NOTE1: if you look at the example that we have in the unit test, which documents the experiment we did with a real vCenter, you can find this exact configuration.
NOTE2: For both the content library and OVA the code will create corresponding K8s objects to make the vm-operator happy, no matter if the content library and OVA already existed before of was created.
NOTE3: might be that in future we should change the config to handle more than one OVA, but let's keep it simple for now

File: test/infrastructure/vcsim/controllers/vspherevm_controller.go [hunk 35]
```diff
@@ -0,0 +1,306 @@
...
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterNameLabel))
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Return early if the object or Cluster is paused.
+	// NOTE: we are using vSphereMachine because VSphereVM is not a valid v1.Object
+	if annotations.IsPaused(cluster, vSphereMachine) {
```
Comment (sbueringer) [inline test/infrastructure/vcsim/controllers/vspherevm_controller.go:125]:
```suggestion
	if annotations.IsPaused(cluster, vSphereVM) {
```
Shouldn't this work?

Maybe I'm misinterpreting the comment

**Comment by sbueringer** [comment]
Last one from my side

File: test/infrastructure/vcsim/controllers/vspherevm_controller.go (hunk 35 above)
Comment (fabriziopandini) [reply test/infrastructure/vcsim/controllers/vspherevm_controller.go]: Fixed, nice catch

**Comment by sbueringer** [comment]
Thank you very much. 

Really nice work :)

/lgtm
/approve
